docs/design/solution-architecture.md

```diff
@@ -0,0 +1,14 @@
+
+A brief summary of Apex operation in this environment is as follows. Nodes N1 and N2 are running the Apex agent and register themselves with the Apex controller SaaS running at public Internet endpoint https://apexcloud.com. In this particular scenario, there is also a relay node (interchangeably also referred to as a "hub router" in the Apex documentation), which is any Linux Virtual machine running in the same Apex Zone as nodes N1 and N2 and directly reachable on the public internet.
+
+The agents determine their own respective public IP address and port via one of two techniques. (1) They initiate a query using the STUN protocol (Session Traversal Utilities for NAT) towards 1 or more well known public STUN servers run by providers such as Google. These servers provide a free service 
+that returns a STUN response containing in its inband payload, the SNAT'ed public IP and source port information it sees for the STUN request, thereby enabling the Apex agents in this case to determine their own public (IP address, port). (2)  They can connect to the Apex controller SaaS which in turn can reflect their public (IP address, port) seen by the controller and the relay node to all Apex nodes in the zone, thereby enabling every node to determine the public reachability information of every other node within the Apex zone (subject to any additional administered policies that may filter distribution of reachability information within a zone).  The apex agents can then attempt to initiate Wireguard  tunnel sessions with each other's public IP information. In most common cases, this attempt shoudl succeed since the NAT gateways at each location have already opened up these public facing ports for the outbound connections initiated previously by the Apex agents when connecting to the Apex controller and relay nodes. In the case where the NAT gateways execute a stricter form of Source NATting known as Symmetric NAT, it may not be possible for apex nodes to initiate wireguard tunnels directly with each other, in which case they have logic to revert to sending traffic to the relay node and having it perform an IP routing operation towards the target Apex node. As far as end user application traffic on each node is concerned, it is routed to use the Wireguard virtual interface and get encrypted and tunneled towards the desired destination wireguard interface IP by the wireguard interface driver logic. 
```

Comment:
2nd technique need a bit of change. The hub router needs to join as a first device. Than it become the peer to all the other nodes in that zone. And ituses the wireguard state (wgctl dump) to find out the endpoints and publish it to controller (api server).